cpp/test/sg/fil_test.cu

```diff
@@ -886,4 +889,139 @@ TEST_P(TreeliteThrowSparse8FilTest, Import) { check(); }
 INSTANTIATE_TEST_CASE_P(FilTests, TreeliteThrowSparse8FilTest,
                         testing::ValuesIn(import_throw_sparse8_inputs));
 
+template <typename T>
+__device__ void serial_multi_reduction(T* in, T* out, int set_size,
+                                       int n_sets) {
+  __syncthreads();
+  if (threadIdx.x < set_size) {
+    int reduction_id = threadIdx.x;
+    T sum = 0;
+    for (int set = 0; set < n_sets; ++set)
+      sum += in[reduction_id + set * set_size];
+    out[reduction_id] = sum;
+  }
+  __syncthreads();
+}
+
+// the most threads a block can have
+const int max_threads = 1024;
+
+struct MultiReductionTestParams {
+  int radix, set_size, n_sets;
+};
+
+template <int R, typename T>
+__device__ void test_single_radix(T value, MultiReductionTestParams p,
+                                  int& error) {
+  __shared__ T work[max_threads], correct_result[max_threads];
+  work[threadIdx.x] = value;
+  serial_multi_reduction(work, correct_result, p.set_size, p.n_sets);
+  T sum = multi_reduction<R>(work, p.set_size, p.n_sets);
+  if (threadIdx.x < p.set_size &&
+      1e-4 < fabsf(sum - correct_result[threadIdx.x])) {
```

Comment:
Consider also testing for `T == int`, where the results should exactly match.

Reply:
that happens automatically. This test does not exercise the ints big enough to differ by more than 1.f once converted to float for the comparison. Thus, the test will not have false negatives or false positives.